src/Entity/Form/DeveloperAppFormTrait.php

```diff
@@ -0,0 +1,98 @@
+ * version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ */
+
+namespace Drupal\apigee_edge\Entity\Form;
+
+use Apigee\Edge\Exception\ApiException;
+use Apigee\Edge\Exception\ClientErrorException;
+use Drupal\Core\Entity\EntityTypeInterface;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Utility\Error;
+
+/**
+ * Helper trait that contains developer app (create/edit) form specific tweaks.
+ *
+ * @see \Drupal\apigee_edge\Entity\Form\DeveloperAppCreateForm
+ * @see \Drupal\apigee_edge\Entity\Form\DeveloperAppCreateFormForDeveloper
+ * @see \Drupal\apigee_edge\Entity\Form\DeveloperAppEditForm
+ * @see \Drupal\apigee_edge\Entity\Form\DeveloperAppEditFormForDeveloper
+ */
+trait DeveloperAppFormTrait {
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function appExists(string $name, array $element, FormStateInterface $formState): bool {
+    // Do not validate if app name is not set.
+    if ($name === '') {
+      return FALSE;
+    }
+
+    // We use the developer app controller factory here instead of entity
+    // query to reduce the number API calls. (Entity query may load all
+    // developers to return whether the given developer has an app with
+    // the provided name already.)
+    /** @var \Drupal\apigee_edge\Entity\Controller\DeveloperAppControllerFactoryInterface $factory */
+    $factory = \Drupal::service('apigee_edge.controller.developer_app_controller_factory');
+    $app = TRUE;
+    try {
+      $app = $factory->developerAppController($formState->getValue('owner'))->load($name);
+    }
+    catch (ApiException $exception) {
+      if ($exception instanceof ClientErrorException && $exception->getEdgeErrorCode() === 'developer.service.AppDoesNotExist') {
+        $app = FALSE;
+      }
+      else {
+        // Fail safe, return TRUE in case of an API communication error or an
+        // unexpected response.
+        $context = [
+          '%app_name' => $name,
+          '%owner' => $formState->getValue('owner'),
+        ];
+        $context += Error::decodeException($exception);
+        \Drupal::logger('apigee_edge')->error("Unable to properly validate an app name's uniqueness. App name: %app_name. Owner: %owner. @message %function (line %line of %file). <pre>@backtrace_string</pre>", $context);
+      }
+    }
+
+    return (bool) $app;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function appEntityDefinition(): EntityTypeInterface {
+    return \Drupal::entityTypeManager()->getDefinition('developer_app');
```

Comment:
Can you inject the entity type manager here? E.g. with an abstract method.

Reply:
I can and I started to implement it a way like that but then I realised creating n+1 abstract method definitions, keeping them in sync and then resolving them in classes that uses this trait seems less maintainable. "Luckily" Drupal 8 provides these beautiful static methods to access services. So it seems to be a fair trade for better maintainability.

As a fallback solution, if we really would like to mock the entity type manager somewere an $entityTypeManager property can be introduced in this trait (but then its visibility must be the same as in all classes that uses this trait) and the could above can be altered like `($this->entityTypeManager ?? \Drupal::entityTypeManager())->getDefinition('developer_app');`